File: Decoria/src/main/java/com/example/Decoria/entity/Voucher.java
package com.example.Decoria.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.GenericGenerator;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "vouchers")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Voucher {

    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    @Column(columnDefinition = "BINARY(16)")
    private UUID id;

    @Column(unique = true, length = 50, nullable = false)
    private String code;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private DiscountType discountType;

    @Column(nullable = false)
    private BigDecimal discountValue;

    private BigDecimal minOrderValue;

    private Integer usageLimit;

    private Integer usedCount = 0;

    @Column(nullable = false)
    private LocalDateTime startDate;

    @Column(nullable = false)
    private LocalDateTime endDate;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    private VoucherStatus status = VoucherStatus.ACTIVE;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    // ⚡ Quan hệ với UserVoucher
    @OneToMany(mappedBy = "voucher")
    private List<UserVoucher> userVouchers;

    // ⚡ Quan hệ với Order
    @OneToMany(mappedBy = "voucher")
    private List<Order> orders;

    public enum DiscountType { PERCENT, FIXED }
    public enum VoucherStatus { ACTIVE, INACTIVE }
}

File: fontend/src/types/Voucher.ts
export type DiscountType = "PERCENT" | "FIXED";
export type VoucherStatus = "ACTIVE" | "INACTIVE";
export type UserVoucherStatus = "SAVED" | "USED"  | "EXPIRED";

export interface IVoucher {
  id: string;
  code: string;
  description?: string;
  discountType: DiscountType;
  discountValue: number;
  minOrderValue: number | null;
  usageLimit: number | null;
  usedCount: number;
  startDate: string;
  endDate: string;
  status: VoucherStatus;
  createdAt: string;
}

export interface CreateVoucherDTO {
  code: string;
  description?: string;
  discountType: DiscountType;
  discountValue: number;
  minOrderValue?: number | null;
  usageLimit?: number | null;
  startDate: string;
  endDate: string;
  // status: VoucherStatus;
}

export interface IUserVoucher {
  id: string;
  voucherId: string;
  code: string;
  description?: string;
  discountType: DiscountType;
  discountValue: number;
  minOrderValue: number | null;
  savedAt: string;
  status: UserVoucherStatus;
}


export type UpdateVoucherDTO = Partial<CreateVoucherDTO>;

export interface ApplyVoucherRequest {
  userId: string;
  orderTotal: number;
  voucherCode: string;
}

export interface ApplyVoucherResponse {
  discount: number;
  finalPrice: number;
  message: string;
  voucherCode: string;
  voucherId: string;
  discountType: DiscountType
  discountValue: number;
}

